core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -137,24 +137,25 @@ class LogSegment(val log: FileMessageSet,
       return FetchDataInfo(offsetMetadata, MessageSet.Empty)
 
     // calculate the length of the message set to read based on whether or not they gave us a maxOffset
-    val length =
-      maxOffset match {
-        case None =>
-          // no max offset, just read until the max position
-          min((maxPosition - startPosition.position).toInt, maxSize)
-        case Some(offset) => {
-          // there is a max offset, translate it to a file position and use that to calculate the max read size
-          if(offset < startOffset)
-            throw new IllegalArgumentException("Attempt to read with a maximum offset (%d) less than the start offset (%d).".format(offset, startOffset))
-          val mapping = translateOffset(offset, startPosition.position)
-          val endPosition =
-            if(mapping == null)
-              logSize // the max offset is off the end of the log, use the end of the file
-            else
-              mapping.position
-          min(min(maxPosition, endPosition) - startPosition.position, maxSize).toInt
-        }
-      }
+    val length = maxOffset match {
+      case None =>
+        // no max offset, just read until the max position
+        min((maxPosition - startPosition.position).toInt, maxSize)
+      case Some(offset) =>
+        // there is a max offset, translate it to a file position and use that to calculate the max read size;
+        // if the max offset is smaller than the start offset, return empty response as it can happen when the max offset
+        // is bounded by the high watermark for a fetch request from normal consumer.
```

Comment:
How about replace the comments in the above 2 lines with the following?

When the leader of a partition changes, it's possible for the new leader's high watermark to be less than the true high watermark in the previous leader for a short window. In this window, if a consumer fetches on an offset between new leader's high watermark and the log end offset, we want to return an empty response.